### Importing Library:

In [1]:
import pandas as pd

### Loading "Matches" Dataset:

In [2]:
data1 = pd.read_csv("matches.csv")
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
data1.head()

id  season       city        date                        team1  \
0   1    2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06               Mumbai Indians   
2   3    2017     Rajkot  2017-04-07                Gujarat Lions   
3   4    2017     Indore  2017-04-08       Rising Pune Supergiant   
4   5    2017  Bangalore  2017-04-08  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

### Filtering KKR and RR matches only:

In [3]:
kkr_rr = data1.loc[((data1['team1'] == 'Rajasthan Royals') & (data1['team2'] == 'Kolkata Knight Riders') | (data1['team1'] == 'Kolkata Knight Riders') & (data1['team2'] == 'Rajasthan Royals'))]
kkr_rr.head()

id  season       city        date                  team1  \
77    78    2008     Jaipur  2008-05-01       Rajasthan Royals   
104  105    2008    Kolkata  2008-05-20  Kolkata Knight Riders   
125  126    2009  Cape Town  2009-04-23       Rajasthan Royals   
167  168    2009     Durban  2009-05-20       Rajasthan Royals   
186  187    2010  Ahmedabad  2010-03-20       Rajasthan Royals   

                     team2            toss_winner toss_decision  result  \
77   Kolkata Knight Riders       Rajasthan Royals           bat  normal   
104       Rajasthan Royals       Rajasthan Royals         field  normal   
125  Kolkata Knight Riders  Kolkata Knight Riders         field     tie   
167  Kolkata Knight Riders  Kolkata Knight Riders         field  normal   
186  Kolkata Knight Riders       Rajasthan Royals           bat  normal   

     dl_applied                 winner  win_by_runs  win_by_wickets  \
77            0       Rajasthan Royals           45               0   
104           0       Rajasthan Royals            0               6   
125           0       Rajasthan Royals            0               0   
167           0  Kolkata Knight Riders            0               4   
186           0       Rajasthan Royals           34               0   

     player_of_match                         venue      umpire1  \
77       SA Asnodkar        Sawai Mansingh Stadium  RE Koertzen   
104        YK Pathan                  Eden Gardens   BG Jerling   
125        YK Pathan                      Newlands    MR Benson   
167        LR Shukla                     Kingsmead   BG Jerling   
186  AA Jhunjhunwala  Sardar Patel Stadium, Motera  RE Koertzen   

            umpire2 umpire3  
77   GA Pratapkumar     NaN  
104     RE Koertzen     NaN  
125       M Erasmus     NaN  
167      SJA Taufel     NaN  
186       RB Tiffin     NaN

## Question 1: Who will win the match?

In [5]:
kkr_rr['winner'].value_counts()

Rajasthan Royals         10
Kolkata Knight Riders    10
Name: winner, dtype: int64

## Question 2:

### Loading "Deliveries" dataset:

In [6]:
data2 = pd.read_csv("deliveries.csv")
data2.head()

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
2         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
3         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
4         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  wide_runs  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0          0         0   
1     2  DA Warner    S Dhawan  TS Mills              0          0         0   
2     3  DA Warner    S Dhawan  TS Mills              0          0         0   
3     4  DA Warner    S Dhawan  TS Mills              0          0         0   
4     5  DA Warner    S Dhawan  TS Mills              0          2         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   
2            0            0             0             4           0   
3            0            0             0             0           0   
4            0            0             0             0           2   

   total_runs player_dismissed dismissal_kind fielder  
0           0              NaN            NaN     NaN  
1           0              NaN            NaN     NaN  
2           4              NaN            NaN     NaN  
3           0              NaN            NaN     NaN  
4           2              NaN            NaN     NaN

In [17]:
wickets = data2.loc[((data2['batting_team'] == "Rajasthan Royals") & (data2['bowling_team'] == "Kolkata Knight Riders") | (data2['batting_team'] == "Kolkata Knight Riders") & (data2['bowling_team'] == "Rajasthan Royals"))]
wickets = pd.DataFrame(wickets.groupby('match_id')['player_dismissed'].count())
wickets

player_dismissed
match_id                  
78                      17
105                     12
126                     15
168                     15
187                     12
227                     11
246                      5
251                     12
314                     15
322                     10
389                     16
427                      8
476                     15
482                     12
570                     15
7908                    11
7942                    14
7951                    11
11312                    5
11334                   13

In [18]:
wickets['player_dismissed'].mean()

12.2

## Question 3:  

In [9]:
abd = data2.loc[data2['batsman'] == 'AB de Villiers']
abd_balls = pd.DataFrame(abd.groupby('match_id')['ball'].count())
abd_runs = pd.DataFrame(abd.groupby('match_id')['batsman_runs'].sum())
abd_data = pd.merge(abd_balls, abd_runs, on = 'match_id')
abd['Total_runs'] = abd['batsman_runs'].expanding().sum()
abd['Total_balls'] = abd['ball'].expanding().count()
abd.loc[abd['Total_runs'] == 30, ['Total_balls', 'Total_runs'] ]
abd_data

F:\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """
F:\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


ball  batsman_runs
match_id                    
8           49            89
12          22            19
17          32            29
27           6             8
30          12             5
33           8             3
37          28            43
42           6            10
45           8            10
79          22            26
82          17            21
87           1             0
93          23            20
95          12             7
104         25            21
125         54           105
129         23            21
133         41            50
136          9             5
141          1             0
150         38            50
154         33            40
158         36            44
161         12            11
165         55            79
167         29            28
170          7             6
172         21            26
177         15             7
180         26            15
183         11            11
185          3             2
189         26            27
197         35            45
221          7             4
237         41            54
242         37            38
245          7             0
248         45            65
263          2             5
268         20            26
280         14            27
291         14            13
296         27            34
302          3             0
304         10            11
306         16            21
307         12            18
312         42            64
317         10            10
320          3             4
326         10             6
329         14            33
333         41            52
338         24            59
344          1             0
350         13            17
356         17            47
362          4             9
367         13            14
376         11             4
392         22            22
396         33            64
401         36            52
407          8             7
411          9            31
417          3             2
420         13            21
426         23            50
429          2             6
432         15            15
438         17            32
441         22            28
448         20            38
452          3             5
462         48            45
468         12            11
471          1             0
475         15            17
481         43            89
484          7             9
488         28            53
492         32            58
495         17            33
499         15            28
503         18            29
506         13            13
510          6            10
522         13            28
525         29            46
534         12            41
538         10            14
540         34            47
546         47            57
550          4             2
554         14            21
557         24            47
562         61           133
566         10            10
568          1             0
574         40            66
575          3             1
580         44            82
587         33            55
590         22            29
592         46            83
595         16            20
603         33            47
606          6             4
611          3             1
615         35            64
617         27            24
620         53           129
624         32            59
626          2             0
632          9             6
633         48            79
636          6             5
7896        23            44
7901        40            58
7904        18            21
7907         2             1
7912        40            92
7917        31            69
7928         4             1
7932         9             6
7938        37            72
7944        41            71
7946        36            53
11137       10             9
11143       42            73
11147        2             1
11150        9            13
11153       34            65
11311       17            18
11319       38            59
11322   

## Question 4:  What will be difference in Eoin Morgan's overall strike rate and the strike rate of the last 7 balls he plays?

#### Filtering data for Eoin Morgan as batsman and calculating overall strike rate:

In [22]:
morgan = data2.loc[(data2['batsman'] == 'EJG Morgan')]
morgan_balls = pd.DataFrame(morgan.groupby('match_id')['ball'].count())
morgan_runs = pd.DataFrame(morgan.groupby('match_id')['batsman_runs'].sum())
morgan_data = pd.merge(morgan_balls, morgan_runs, on='match_id')
morgan_data['Strike Rate'] = 100 * morgan_data['batsman_runs']/morgan_data['ball']
morgan_data = morgan_data.rename({'ball' : 'Balls', 'batsman_runs' : 'Runs'}, axis = 1)
morgan_data = morgan_data.rename({'Strike Rate': 'Overall SR'}, axis=1)
morgan_data

Balls  Runs  Overall SR
match_id                         
15           20    22  110.000000
19           17    13   76.470588
32           21    26  123.809524
54            4     4  100.000000
178           7    10  142.857143
188           6     7  116.666667
192           4     1   25.000000
197          14    17  121.428571
235           6     6  100.000000
255          12    10   83.333333
258           6     6  100.000000
266           1     0    0.000000
270          15    28  186.666667
275          18    14   77.777778
278          52    66  126.923077
281           8     5   62.500000
291           5     2   40.000000
382          16    14   87.500000
389          39    51  130.769231
392           5     2   40.000000
397          24    47  195.833333
400          39    47  120.512821
406           7     2   28.571429
413          25    31  124.000000
415          27    42  155.555556
418          25    32  128.000000
424           9    10  111.111111
434          10     5   50.000000
437          15    15  100.000000
453           9     9  100.000000
529          30    27   90.000000
531           2     1   50.000000
551          27    32  118.518519
555          11     5   45.454545
558          30    63  210.000000
561          25    22   88.000000
564           7    17  242.857143
568           7    11  157.142857
572          15     9   60.000000
580          18    22  122.222222
584          43    51  118.604651
588          12    11   91.666667
594          20    25  125.000000
598           2     0    0.000000
628          10    14  140.000000

#### Filtering data for Eoin Morgan as batsman and the last 7 ball he plays:

In [20]:
morgan2 = data2.loc[(data2['batsman'] == 'EJG Morgan') & ((data2['over'] == 18) & (data2['ball'] == 5) | ((data2['over'] >= 19) & (data2['over'] <= 20)))]
morgan2_balls = pd.DataFrame(morgan2.groupby('match_id')['ball'].count())
morgan2_runs = pd.DataFrame(morgan2.groupby('match_id')['batsman_runs'].sum())
morgan2_data = pd.merge(morgan2_balls, morgan2_runs, on='match_id')
morgan2_data['Strike Rate'] = 100 * morgan2_data['batsman_runs']/morgan2_data['ball']
morgan2_data = morgan2_data.rename({'ball' : 'Balls', 'batsman_runs' : 'Runs'}, axis = 1)
morgan2_data = morgan2_data.rename({'Strike Rate': 'Last 7 Ball SR'}, axis=1)
morgan2_data

Balls  Runs  Last 7 Ball SR
match_id                             
188           3     3      100.000000
258           5     5      100.000000
278           3     2       66.666667
382           3     2       66.666667
389           2     1       50.000000
397           4     7      175.000000
413           1     4      400.000000
415           1     6      600.000000
418          10     9       90.000000
551           8    16      200.000000
580           5     5      100.000000
584           2     1       50.000000

#### Merging above 2 Dataframes:

In [28]:
sr_diff = pd.merge(morgan_data, morgan2_data, on = 'match_id')
sr_diff = sr_diff[['Overall SR', 'Last 7 Ball SR']]
sr_diff['SR Difference'] = (sr_diff['Overall SR'] - sr_diff['Last 7 Ball SR'])
sr_diff

Overall SR  Last 7 Ball SR  SR Difference
match_id                                           
188       116.666667      100.000000      16.666667
258       100.000000      100.000000       0.000000
278       126.923077       66.666667      60.256410
382        87.500000       66.666667      20.833333
389       130.769231       50.000000      80.769231
397       195.833333      175.000000      20.833333
413       124.000000      400.000000    -276.000000
415       155.555556      600.000000    -444.444444
418       128.000000       90.000000      38.000000
551       118.518519      200.000000     -81.481481
580       122.222222      100.000000      22.222222
584       118.604651       50.000000      68.604651

#### Average Strike Rate difference:

In [27]:
diff = abs(sr_diff['Overall SR'].mean() - sr_diff['Last 7 Ball SR'].mean())
diff

39.47833984849491

**Answer:** So, we found that average difference between Morgan's Overall Strike Rate and Last 7 ball strike rate is around **39.50**.

## Question 5: How many runs will RCB score off KKR in the death?

#### Solution: Filtering data with RCB and KKR Matches where batting team is RCB and bowling team is KKR:

In [43]:
rcb_runs = data2.loc[(data2['batting_team'] == "Royal Challengers Bangalore") & (data2['bowling_team'] == "Kolkata Knight Riders") & ((data2['over'] >= 16) & (data2['over'] <= 20))]
rcb_runs = pd.DataFrame(rcb_runs.groupby('match_id')['total_runs'].sum())
rcb_runs

total_runs
match_id            
45                53
60                 1
88                13
134               45
156               57
178               46
217               42
258               23
317               42
344               56
392               41
441               31
468               32
506               59
522               60
606               73
624               51
7896              45
7922              69
11153             66
11326             94

#### Getting mean of the death over runs match wise:

In [44]:
avg_runs = rcb_runs['total_runs'].mean()
avg_runs

47.57142857142857

**Answer:** Average death over runs in each match is 47.57.